test/e2e/apimachinery/flowcontrol.go

```diff
@@ -135,3 +299,56 @@ func testResponseHeaderMatches(f *framework.Framework, impersonatingUser, plUID,
 	}
 	return true
 }
+
+// uniformQPSLoadSingle loads the API server with requests at a uniform <qps>
+// for <loadDuration> time. The number of successfully completed requests is
+// returned.
+func uniformQPSLoadSingle(f *framework.Framework, username string, qps float64, loadDuration time.Duration) int32 {
+	var completed int32
+	var wg sync.WaitGroup
+	ticker := time.NewTicker(time.Duration(1e9/qps) * time.Nanosecond)
+	defer ticker.Stop()
+	timer := time.NewTimer(loadDuration)
+	for {
+		select {
+		case <-ticker.C:
+			wg.Add(1)
+			// Each request will have a non-zero latency. In addition, there may be
+			// multiple concurrent requests in-flight. As a result, a request may
+			// take longer than the time between two different consecutive ticks
+			// regardless of whether a requests is accepted or rejected. For example,
+			// in cases with clients making requests far above their concurrency
+			// share, with little time between consecutive requests, due to limited
+			// concurrency, newer requests will be enqueued until older ones
+			// complete. Hence the synchronisation with sync.WaitGroup.
+			go func() {
+				defer wg.Done()
+				makeRequest(f, username)
+				atomic.AddInt32(&completed, 1)
+			}()
+		case <-timer.C:
+			// Still in-flight requests should not contribute to the completed count.
+			totalCompleted := atomic.LoadInt32(&completed)
```

Comment:
Hm, this gives requests sent very soon before the time limit a different amount of time to complete in than requests sent early on.

Reply:
Maybe this is OK for this test. Other approaches also have downsides.

Reply:
I don't think this makes much of a difference -- the only requests that would get missed are the ones that, as you said, get fired just before the timer ends. I guess the whole thing depends on how we define a "completed" request -- if we wait for such requests to finish, are we testing the system under load? For example, the high QPS client may have completed, so the low QPS client's requests would easily go through, but it shouldn't really be counted because there was no request drown out.